design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+`KsqlTableFunction` which is a wrapper around the actual table function (whether built-in or
+generated). We'll also need to augment the function registry to manage the addition and retrieval
+of table functions.
+* Augment the query analysis code to extract information relating to any table functions
+in the query. We can do this either by extending the existing query analyzer or perhaps creating
+a new `TableFunctionAnalyzer` in a similar way to how aggregate analysis uses its own 
+`AggregateAnalyzer`.
+* Augment the building of logical plan by building a new node representing a FlatMap stage of the
+stream processing in the case there are table functions in the SELECT.
+* Augment the execution plan logic by adding new nodes in the execution plan corresponding to any
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
+The number of rows returned from the FlatMap will be equal to the greatest of the number of rows
+returned from any of the table functions for that row. Any non table function values for the row will
+be duplicated for each row. For any table function that returned fewer than the greatest number of rows
+the values for that columnn will be `null`. For performance reasons, the FlatMap logic should cache any
+ table functions for the columns to avoid them being looked up each time.
+
+### Implement set of built in table functions
+
+We will implement the following built in table functions:
+
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `EXPLODE(MAP<K, V>)` -> Results in `N` rows (where `N` is number of elements in map) each with a column
+holding a `STRUCT<key K, value V>` that contains the key and value.
+* `INDEX_EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a
+column of type `STRUCT<index INT, value V>` containing the index of the element from the original array
+and the value from the array.
+
+### Implement framework of user defined table functions
+
+We will introduce annotations `@UdtfFactory` and `@UdtfDescription` in a similar way to UDAFs.
+
+This will allow users to implement their own table functions which can be used in queries.
+
+We will implement the loading of any UDTFs from the classpath at startup of the server.
+
+## Test plan
+
+New unit/module level tests will be written as appropriate.
+
+A set of new QueryTranslationTest tests will be written to cover different ways of using table
+functions in queries. These will cover all the built-in functions as well as testing examples of user defined table
+functions.
+
+## Documentation Updates
+
+The following changes to the documentation will be made:
```

Comment:
This docs section needs expanding as the requirements/approach solidify.  Ideally, this docs section should _be_ the text that will ultimately end up in the real docs.  Adding the documentation here allows those reviewing the KLIP to see the same docs the user will ultimately see.  If we can't understand if from the docs then either the docs ain't right, or the solution doesn't hang together.